mosqagent.c

```diff
@@ -105,6 +107,100 @@ void mqtta_dispose_message(struct mqtta_message *msg)
     free(msg);
 }
 
+/**
+ * If there is already a configuration in the agent, make sure to take care
+ * of the memory object before calling this function, otherwise there may be
+ * memory leaks!
+ */
+int mqtta_load_configuration(struct mosqagent *agent,
+                             const char* filepath)
+{
+    int ret = 0;
+
+    struct mosqagent_config *config;
+
+    config_t configuration;
+
+    const char* client_name;
+    const char* broker_host;
+
+    // Init the configuration struct from libconfig
+    config_init(&configuration);
+
+    /*
+     * Check this part first. Not finding the config file is the most likely
+     * failure and we'll fail this before reserving memory.
+     */
+
+    if (!config_read_file(&configuration, filepath)) {
+        fprintf(stderr, "Cannot read config file: %s\n", config_error_text(&configuration));
+        ret = 1;
+        goto cleanup_with_configuration;
+    }
+
+    // Now create the memory object
+    config = malloc(sizeof(*config));
+    if (!config) {
+        ret = -ENOMEM;
+        goto cleanup_with_configuration;
+    }
+
+    // We have to find an agent name!
+    if (config_lookup_string(&configuration, "mosqagent.name", &client_name))
+    {
+        config->client_name = malloc(strlen(client_name));
+        strcpy(config->client_name, client_name);
+    } else {
+        printf("Cannot retrieve client name!");
+        ret = 2;
+        goto fail_with_config_object;
+    }
+
+    // Host is optional
+    if (config_lookup_string(&configuration, "mosqagent.broker.host", &broker_host))
+    {
+        config->host = malloc(strlen(broker_host));
+        strcpy(config->host, broker_host);
+    } else {
+        printf("Host not found, using configuration default.");
+    }
+
+    // Port is optional
+    config_lookup_int(&configuration, "mosqagent.broker.port", &config->port);
+
+    // If we got through to here, store configuration to agent
+    agent->config = config;
+
+    // standard cleanup
+    goto cleanup_with_configuration;
+
+
+    fail_with_config_object:
+    free(config);
```

Comment:
Indent jump labels one level less, the same as the `{` and `}` of the surrounding block, or let's say don't indent them at all. They should stand out optically.